src/backend/test.rs

```diff
@@ -140,6 +198,32 @@ impl fmt::Display for TestBackend {
     }
 }
 
+/// Append the provided cells to the bottom of a scrollback buffer. The number of cells must be a
+/// multiple of the buffer's width. If the scrollback buffer ends up larger than 65535 lines tall,
+/// then lines will be removed from the top to get it down to size.
+fn append_to_scrollback(scrollback: &mut Buffer, cells: impl IntoIterator<Item = Cell>) {
+    let width = scrollback.area.width as usize;
+    let len_before = scrollback.content.len();
+    scrollback.content.extend(cells);
+    let cells_appended = scrollback.content.len() - len_before;
+    let lines_appended = cells_appended / width;
+    assert_eq!(
+        cells_appended,
+        lines_appended * width,
+        "count of appended cells not a multiple of scrollback buffer width"
+    );
+
+    let new_scrollback_height = scrollback.area.height as usize + lines_appended;
+    if u16::try_from(new_scrollback_height).is_ok() {
+        scrollback.area.height = new_scrollback_height as u16;
+    } else {
+        scrollback
+            .content
+            .drain(0..width * (new_scrollback_height - u16::MAX as usize));
+        scrollback.area.height = u16::MAX;
+    }
+}
```

Comment:
This code specifically allows the scrollback buffer to be larger than `u16::MAX` `Cell`s. That fact is actually tested in the test for overflowing scrollback.

Here I'm just trying to limit the scrollback buffer size to a certain height. The max height could have been anything that fits inside of a `u16`. I just chose `u16::MAX` because it's convenient.

In #1286 you discuss how this behavior of certain `Buffer` functions assuming the area is clamped to `u16::MAX` may not be desirable. I was trying to write code that didn't suffer from this limitation and also didn't rely on this behavior.

Reply:
Ah got it. Makes sense.

Does this do the same thing as the code?:
```rust
    scrollback.content.extend(cells);
    let width = scrollback.area.width as usize;
    let new_height = (scrollback.content.len() / width).min(u16::MAX as usize);
    let keep_from = scrollback.content.len().saturating_sub(width * u16::MAX as usize);
    scrollback.content.drain(0..keep_from);
    scrollback.area.height = new_height as u16;
```

I generally avoid putting asserts inside functions like this. (as they make for difficult to debug issues for end users, and this should never be hit).

Reply:
Yeah, that code is equivalent. You don't need the saturating_sub, since keep_from will always be >= 0. I don't know what assert you're talking about. Are you saying you're using saturating_sub instead of vanilla - or checked_sub because the latter two effectively have asserts? If so, that's the opposite of how I usually write code, but I'll go with the flow. When I see saturating_sub, it usually indicates to me that it could saturate, whereas the other two prove to me that it won't ever. But again, I'm happy to go with the flow.